src/System.Text.Json/src/System/Text/Json/Serialization/JsonPropertyInfoCommon.cs

```diff
@@ -69,19 +80,52 @@ internal override object GetValueAsObject(object obj, JsonSerializerOptions opti
                 return obj;
             }
 
-            Debug.Assert(Get != null);
+            Debug.Assert(HasGetter);
             return Get((TClass)obj);
         }
 
         internal override void SetValueAsObject(object obj, object value, JsonSerializerOptions options)
         {
-            Debug.Assert(Set != null);
+            Debug.Assert(HasSetter);
             TDeclaredProperty typedValue = (TDeclaredProperty)value;
 
             if (typedValue != null || !IgnoreNullPropertyValueOnWrite(options))
             {
-                Set((TClass)obj, (TDeclaredProperty)value);
+                Set(obj, typedValue);
             }
         }
+
+        internal TDeclaredProperty Get(object obj)
+        {
+            if (default(TClass) == null)
+            {
+                var get = (GetPropertyValue)_get;
+                return get((TClass)obj);
+            }
+            else
+            {
+                var get = (GetPropertyValueByRef)_get;
+                return get(ref Unsafe.As<ByReference>(obj).Value);
```

Comment:
Doesn't this only work for your test class with a single value?

Reply:
`ByReference` represents a boxed object, this is how the CLR stores value types on the heap. The code does unsafe cast of a boxed value to `ByReference` and gets a reference to its actual value.

Reply:
@jkotas any concern using class like this to get access to the real value from a boxed object?:
```
private sealed class ByReference 
{ 
        public TClass Value; 
} 

...
ref Unsafe.As<ByReference>(obj).Value
```

Reply:
This is unsafe construct that depends on runtime implementation details. I do not think it is a good idea to use it in high-level libraries like this one.

Can you use `Unsafe.Unbox` instead?

Reply:
Somehow missed this method, thanks for pointing out. It will make the code a bit trickier, but it's definitely possible to use `Unsafe.Unbox<T>` here.